File: WebContent/js/websocket.js
/*
Snake vs Snake - An Snake multiplayer game.
Copyright (C) 2016  Davidson Francis <davidsondfgl@gmail.com>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as
published by the Free Software Foundation, either version 3 of the
License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>. 
*/
var socket;
var registered = false;

function startClient() 
{
	var wsUri;
	console.log("opening socket");
	
	/* Localhost. */
	if (document.domain == "127.0.0.1")
		wsUri = "ws://127.0.0.1:8080/snake/game";
	
	/* My private server. */
	else if (document.domain == "tlds.ddns.net")
		wsUri = "ws://" + document.location.hostname + ":8080/snake/game";
	
	/* OpenShift with Wildfly. */
	else
		wsUri = "ws://" + document.location.hostname + ":8000"+ document.location.pathname + "/../game";

	socket = new WebSocket(wsUri);

	/* When the connection is opened, login. */
	socket.onopen = function()
	{
		console.log("Opened socket.");
		var nickname = $("#txtName").val();
		socket.send("connect|" + nickname);
		$(document).keydown(getSnakeMovement);
	};

	/* When received a message, parse it and either add/remove user or post message. */
	socket.onmessage = function(a)
	{
		/* Process the message here. */
		console.log("received message: " + a.data);
		var message = a.data.split("|");	
		
		switch(message[0])
		{
			case "yourSnake":
				addSnake(message[1]);
				break;
				
			case "tryAnotherName":
				displayMessage("red", "There is already someone with that name, please try another!", 5000, "top");
				break;
				
			case "serverFull":		
				displayMessage("red", "The server is full, try again later! :\'(", 5000, "top");
				break;
				
			case "newPlayer":
				var nick = message[1];
				var color = JSON.parse(message[2]).snake.colorBody;
				displayMessage("green", "~~ <b><font color=\"" + color + "\">" + nick + "</font></b> entered in the world! ~~", 5000, "top");
				newPlayer(message[2]);
				break;
				
			case "removePlayer":
				var nick = message[1];
				displayMessage("red", "~~ <b>" + nick + "</b> left the world! ~~", 5000, "top");
				removePlayer(nick);
				break;
				
			case "updatePlayer":
				updatePlayer(message[1]);
				break;
				
			case "hello":
				addPlayer(message[1]);
				break;
				
			case "newApple":
				updateApple(message[1]);
				break;
				
			case "endGame":
				announceWinner(message[1]);
				socket.close();
				break;
		}
	};
			
	socket.onclose = function()
	{
		console.write("Closed socket.");
	};
		
	socket.onerror = function()
	{
		console.write("Error during transfer.");
	};
}

function displayMessage(color, msg, time, align)
{
	var type;
	
	switch(color)
	{
		case "red":
			type = "error";
			break;
			
		case "yellow":
			type = "warning";
			break;
			
		case "green":
			type = "success";
			break;
			
		case "blue":
			type = "information";
			break;
			
		default:
			type = "success";
			break;
	}
	
	if (align == "top")
	{
		var noty_id = noty({
			layout: 'topCenter',
			type: type,
	   		text: msg,
	   		theme: 'relax',
			timeOut: 500,
		});
	    
		setTimeout(function(){
	  		$(".noty_message").click();
		}, 500+time);
	}
	else if (align == "center")
	{
		var noty_id = noty({
			layout: 'center',
			type: type,
	   		text: msg,
	   		theme: 'relax',
			timeOut: 500,
		});
	    
		setTimeout(function(){
	  		$(".noty_message").click();
		}, 500+time);
	}
}

function sendMessage(msg)
{
	if (msg != null && msg != "")
		socket.send(msg);
}

$(document).ready(function()
{
	$('#btnPlay').click(function()
	{
		if ($("#txtName").val())
			startClient();
		else
			alert("Plz, type your nickname!");
	});
});

File: src/lt/wc/pyl/WebSocket.java
/*
Snake vs Snake - An Snake multiplayer game.
Copyright (C) 2016  Davidson Francis <davidsondfgl@gmail.com>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as
published by the Free Software Foundation, either version 3 of the
License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>. 
*/
package lt.wc.pyl;

import java.io.IOException;
import java.util.HashSet;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;

import javax.websocket.CloseReason;
import javax.websocket.OnClose;
import javax.websocket.OnError;
import javax.websocket.OnMessage;
import javax.websocket.OnOpen;
import javax.websocket.Session;
import javax.websocket.server.ServerEndpoint;

import lt.wc.pyl.Snake.*;
import lt.wc.pyl.Snake.Food.Apple;

@ServerEndpoint("/game")
public class WebSocket
{
	private static Set<Session> conns = java.util.Collections
			.synchronizedSet(new HashSet<Session>());
	
	private static Map<Session, String> playerSession = new ConcurrentHashMap<Session, String>();
	private static Map<String, Snake> playerSnake = new ConcurrentHashMap<String, Snake>();
	private static Apple apple = null;
	
	private Session currentSession;

	public WebSocket()
	{
		Logging.println("WebSocket Constructed!");
	}

	@OnOpen
	public void onOpen(Session session)
	{
		Logging.println("WebSocket opened session: " + session.getId());
		conns.add(session);
		this.currentSession = session;
	}

	@OnMessage
	public void onMessage(String message)
	{
		String msg[] = message.split("\\|");
		Logging.println("Received: " + message);
		
		switch(msg[0])
		{
			/* onPlayerConnect. */
			case "connect":
				addPlayer(msg[1]);
				break;
			
			/* onPlayerUpdate. */
			case "updateData":
				Snake s = Snake.getSnake(msg[1]);
				if ((s.pieces.size()-1) != Game.snakeLength)
					broadcastMessage("updatePlayer|" + msg[1], playerSession.get(currentSession));
				else
					broadcastMessage("endGame|" + s.playerName, null);
				break;
			
			/* onEnemyConnect. */
			case "iAM":
				sendMessage("hello|" + msg[2], null, msg[1]);
				break;
			
			/* onEatApple. */
			case "updateApple":
				updateApple();
				break;
		}
	}

	@OnClose
	public void onClose(Session session, CloseReason reason)
	{		
		String nick = playerSession.get(session);
		conns.remove(session);
		playerSession.remove(session);
		
		if (nick != null)
		{
			playerSnake.remove(nick);
			broadcastMessage("removePlayer|" + nick, null);
		}
		
		Logging.println("Closing a WebSocket due to, code: " + reason.getCloseCode() + ", see u later " + nick);
	}

	@OnError
	public void onError(Session session, Throwable throwable)
	{
		String nick = playerSession.get(session);
		conns.remove(session);
		playerSession.remove(session);
		
		if (nick != null)
		{
			playerSnake.remove(nick);
			broadcastMessage("removePlayer|" + nick, null);
		}
		
		Logging.println("An error has ocurred in WebSocket, removing " + nick + ", see u later");
	}
	
	/**
	 * Adds the new user to the server.
	 * @param username Player name
	 */
	private void addPlayer(String username)
	{
		if ( !playerSnake.containsKey(username) && !playerSession.containsKey(currentSession) )
		{
			Snake snake = Game.generateSnake(playerSnake, username);
			if (snake != null)
			{
				/* Sends his snake for everyone. */
				broadcastMessage("newPlayer|" + username + "|" + snake, null);
				
				/* Sends the snake to him. */
				playerSession.put(currentSession, username);
				playerSnake.put(username, snake);
				
				sendMessage("yourSnake|" + snake, currentSession, null);

				Logging.println("User " + username + " entered in the world!");
				Logging.println("json: "+ snake);
				
				/* Generate the first Apple or not */
				if (playerSession.size() == 1 && apple == null)
				{
					apple = Apple.getRandomApple();
					broadcastMessage("newApple|" + apple, null);
					Logging.println("Generating new apple...");
				}
				else
					sendMessage("newApple|" + apple, currentSession, null);
			}
			else
			{
				sendMessage("serverFull", currentSession, null);
				Logging.println("Impossible to get an Snake, probably server is full?");
			}
		}
		else
		{
			sendMessage("tryAnotherName", currentSession, null);
			Logging.println("Attempting to use an existent name: " + username);
		}
	}
	
	/**
	 * Update all the apples of the players.
	 */
	private void updateApple()
	{
		apple = Apple.getRandomApple();
		broadcastMessage("newApple|" + apple, null);
	}
	
	/**
	 * Broadcast an string for all users if ignoreUser is null, otherwise<br>
	 * broadcast to all but him.
	 * @param message Message to send.
	 * @param ignoreUser User to be ignored or null.
	 */
	private void broadcastMessage(String message, String ignoreUser)
	{
		if (ignoreUser != null)
		{
			for (Entry<Session, String> entry : playerSession.entrySet())
			{
				try
				{
					if (!entry.getValue().equals(ignoreUser))
						entry.getKey().getBasicRemote().sendText(message);
				}
				catch (IOException ioe)
				{
					Logging.println("Failed while broadcasting this message: " + message);
				}
			}
		}
		else
		{
			for (Entry<Session, String> entry : playerSession.entrySet())
			{
				try
				{
					entry.getKey().getBasicRemote().sendText(message);
				}
				catch (IOException ioe)
				{
					Logging.println("Failed while broadcasting this message: " + message);
				}
			}
		}
	}
	
	/**
	 * Sends an message by an specific user, if session is not null<br>
	 * sends the message by using session, otherwise, uses the name<br>
	 * provided.
	 * @param message Message to be send.
	 * @param session Users session.
	 * @param name Users name.
	 */
	private void sendMessage(String message, Session session, String name)
	{
		if (session != null)
		{
			try
			{
				session.getBasicRemote().sendText(message);
			}
			catch (IOException ioe)
			{
				Logging.println("Failed while sending this message: " + message);
			}
		}
		else
		{
			for (Entry<Session, String> entry : playerSession.entrySet())
			{
				try
				{
					if (entry.getValue().equals(name))
					{
						entry.getKey().getBasicRemote().sendText(message);
						break;
					}
				}
				catch (IOException ioe)
				{
					Logging.println("Failed while broadcasting this message: " + message);
				}
			}
		}
	}
}
